xla/pjrt/cpu/mpi_collectives.cc

```diff
@@ -0,0 +1,461 @@
+
+namespace xla::cpu {
+
+absl::StatusOr<MPI_Datatype> PrimitiveTypeToMpiType(
+    PrimitiveType element_type) {
+  switch (element_type) {
+    case S8:
+      return MPI_INT8_T;
+    case U8:
+    case PRED:
+      return MPI_UINT8_T;
+    case S16:
+      return MPI_INT16_T;
+    case U16:
+      return MPI_UINT16_T;
+    case S32:
+      return MPI_INT32_T;
+    case U32:
+      return MPI_UINT32_T;
+    case S64:
+      return MPI_INT64_T;
+    case U64:
+      return MPI_UINT64_T;
+    case F32:
+      return MPI_FLOAT;
+    case F64:
+      return MPI_DOUBLE;
+    case C64:
+      return MPI_C_COMPLEX;
+    case C128:
+      return MPI_C_DOUBLE_COMPLEX;
+    default:
+      // For implementing the reduction of unsupported types
+      // see e.g. https://stackoverflow.com/a/29643391
+      return absl::InvalidArgumentError(absl::StrCat(
+          "Unsupported primitive type for reduction: ",
+          primitive_util::LowercasePrimitiveTypeName(element_type)));
+  }
+}
+
+bool MpiTypeIsComplex(MPI_Datatype type) {
+  return type == MPI_C_COMPLEX || type == MPI_C_DOUBLE_COMPLEX;
+}
+
+absl::StatusOr<MPI_Op> ReductionKindToMpiOp(ReductionKind reduction_kind,
+                                            MPI_Datatype type) {
+  switch (reduction_kind) {
+    case ReductionKind::SUM:
+      return MPI_SUM;
+    case ReductionKind::PRODUCT:
+      return MPI_PROD;
+    case ReductionKind::MIN:
+      if (!MpiTypeIsComplex(type)) {
+        return MPI_MIN;
+      } else {
+        return absl::InvalidArgumentError(
+            "MIN reduction not supported for complex types");
+      }
+    case ReductionKind::MAX:
+      if (!MpiTypeIsComplex(type)) {
+        return MPI_MAX;
+      } else {
+        return absl::InvalidArgumentError(
+            "MAX reduction not supported for complex types");
+      }
+    default:
+      return absl::InvalidArgumentError(
+          absl::StrCat("Unknown reduction kind: ", reduction_kind));
+  }
+}
+
+static absl::Status MpiErrorToAbslStatus(int error) {
+  if (error != MPI_SUCCESS) {
+    char error_str[MPI_MAX_ERROR_STRING];
+    int len;
+    MPI_Error_string(error, error_str, &len);
+    return absl::UnknownError(absl::StrCat("MPI error: ", error_str));
+  }
+  return absl::OkStatus();
+}
+
+MpiCollectivesCommunicator::MpiCollectivesCommunicator(
+    std::vector<int> mpi_world_ranks) {
+  MPI_Group group_world, group;
+
+  // Create a MPI communicator conatining the ranks from mpi_world_ranks.
+  // It's MPI ranks respect the order of mpi_world_ranks, which means that the
+  // MPI communicator rank and MpiCollectivesCommunicator rank coincide.
+
+  MPI_Comm_group(MPI_COMM_WORLD, &group_world);
+  MPI_Group_incl(group_world, mpi_world_ranks.size(), mpi_world_ranks.data(),
+                 &group);
+  MPI_Comm_create(MPI_COMM_WORLD, group, &comm_);
+  MPI_Comm_rank(comm_, &mpi_rank_);
+  MPI_Comm_size(comm_, &mpi_size_);
+
+  MPI_Group_free(&group_world);
+  MPI_Group_free(&group);
```

Comment:
I think we can bypass calling `ExchangeGlobalDeviceIds()` altogether using `MPI_Comm_split`.
1. Pass the current device's `global_device_id` as an input parameter to `MpiCollectivesCommunicator::MpiCollectivesCommunicator`. Also pass a bool `in_group` indicating whether the current device is part of this new communication group.
2. Call `MPI_Comm_split`.
 ```suggestion
  MPI_Comm_split(MPI_COMM_WORLD, in_group ? 0 : MPI_UNDEFINED, global_device_id, &comm_);
  MPI_Comm_rank(comm_, &mpi_rank_);
  MPI_Comm_size(comm_, &mpi_size_);
```
Processes that isn't in that group will get `MPI_COMM_NULL` assigned to `comm_`.
What do you think?